### PR TITLE
AHCI: add support for basing quirky controllers on AHCIController + add hacky fix for some intel AHCI chipsets

### PR DESCRIPTION
**Pull request by ElectrodeYT**
CC @supercomputer7
The code for AHCIPortHandler will probably need some changing, as its really weird

### CODE REVIEW

**Comment by ElectrodeYT** [comment]
Made it into a draft as there are still a couple quirks that i need to take care off relating to the Phy

**Comment by supercomputer7** [comment]
Let's try to stick to the formatting of commit messages as explained [here](https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md.).

Also, let's try to create code which looks and use C++ features like virtual functions. I'll put some notes for discussion so you can see what I'm referring to.

File: Kernel/Storage/AHCIController.cpp [hunk 1]
```diff
@@ -79,13 +79,14 @@ volatile AHCI::HBA& AHCIController::hba() const
...
 
-AHCIController::AHCIController(PCI::Address address)
+AHCIController::AHCIController(PCI::Address address, bool init)
     : StorageController()
     , PCI::DeviceController(address)
     , m_hba_region(hba_region())
     , m_capabilities(capabilities())
 {
-    initialize();
+    if (init)
```
Comment (supercomputer7) [inline Kernel/Storage/AHCIController.cpp:88]: Maybe instead of passing a boolean parameter, we could make `initialize` a virtual function so we can override it. In that  case, we will not call initialize method from the constructor because it will use the method from the parent class and not from the derived class.

File: Kernel/Storage/AHCIPort.h [hunk 2]
```diff
@@ -65,36 +66,36 @@ class AHCIPort : public RefCounted<AHCIPort> {
     void rebase();
     void recover_from_fatal_error();
     bool shutdown();
-    ALWAYS_INLINE void spin_up() const;
-    ALWAYS_INLINE void power_on() const;
+    void spin_up() const;
```
Comment (supercomputer7) [inline Kernel/Storage/AHCIPort.h:69]: Is there a good reason to not make this `ALWAYS_INLINE`?

File: Kernel/Storage/AHCIPortHandler.cpp [hunk 3]
```diff
@@ -14,6 +14,43 @@ NonnullRefPtr<AHCIPortHandler> AHCIPortHandler::create(AHCIController& controlle
     return adopt_ref(*new AHCIPortHandler(controller, irq, taken_ports));
 }
 
+NonnullRefPtr<AHCIPortHandler> AHCIPortHandler::create(AHCIController& controller, u8 irq, AHCI::MaskedBitField taken_ports, NonnullRefPtr<AHCIPort> (*ahci_port_creator)(const AHCIPortHandler&, volatile AHCI::PortRegisters&, u32 port_index))
+{
+    return adopt_ref(*new AHCIPortHandler(controller, irq, taken_ports, ahci_port_creator));
+}
+
+AHCIPortHandler::AHCIPortHandler(AHCIController& controller, u8 irq, AHCI::MaskedBitField taken_ports, NonnullRefPtr<AHCIPort> (*ahci_port_creator)(const AHCIPortHandler&, volatile AHCI::PortRegisters&, u32 port_index))
```
Comment (supercomputer7) [inline Kernel/Storage/AHCIPortHandler.cpp:22]: Hmm, I really don't like how this thing works... we essentially pass a pointer (and not a `Function` like we should...) of a function. I'm almost sure we can do it in a better way, and not to pass a callback to some sort of creator.

File: Kernel/Storage/IntelAHCIController.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    return adopt_ref(*new IntelAHCIController(address));
+}
+
+IntelAHCIController::IntelAHCIController(PCI::Address address)
+    : AHCIController(address, false)
+{
+    initialize();
+}
+
+void IntelAHCIController::initialize()
```
Comment (supercomputer7) [inline Kernel/Storage/IntelAHCIController.cpp:25]: This feels like a duplication of `AHCIController::initialize` method. Maybe we can do it in a better way, like do specific stuff for the Intel AHCI controller and then call the more generic `AHCIController::initialize` method?

File: Kernel/Storage/IntelAHCIPort.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+IntelAHCIPort::IntelAHCIPort(const AHCIPortHandler& handler, volatile AHCI::PortRegisters& regs, u32 port_index)
+    : AHCIPort(handler, regs, port_index)
+{
+    PCI::Address controller_address = handler.hba_controller()->pci_address();
+    if(PCI::get_id(controller_address).device_id == 0x3b22) {
+        m_has_reset_quirk = true;
+    }
+}
+
+bool IntelAHCIPort::initialize(ScopedSpinLock<SpinLock<u8>>& main_lock)
```
Comment (supercomputer7) [inline Kernel/Storage/IntelAHCIPort.cpp:29]: This feels like a duplication of `AHCIPort::initialize` method. Maybe we can do it in a better way, like do specific stuff for the Intel AHCI controller ports and then call the more generic `AHCIPort::initialize` method?

**Review by supercomputer7 — COMMENTED** [review]
These are the notes for the technical improvments. Keep up the good job!

**Comment by supercomputer7** [comment]
Hi,
As some time passed, I'd be happy if you could fix the suggestions I put earlier :)
Also, it'd be very nice if you put the reason for this change in a clear & expressed manner. A log file of a crash on a bare metal test machine is a good example of such thing, so we can all understand what actually happened on your side.

**Comment by ElectrodeYT** [comment]
> Hi,
> As some time passed, I'd be happy if you could fix the suggestions I put earlier :)
> Also, it'd be very nice if you put the reason for this change in a clear & expressed manner. A log file of a crash on a bare metal test machine is a good example of such thing, so we can all understand what actually happened on your side.

I'll save you the life story, but basically I can't really work on much of anything until at least July 2.

TLDR: am failing 2 classes, really need to work on school shit

**Comment by supercomputer7** [comment]
> > Hi,
> > As some time passed, I'd be happy if you could fix the suggestions I put earlier :)
> > Also, it'd be very nice if you put the reason for this change in a clear & expressed manner. A log file of a crash on a bare metal test machine is a good example of such thing, so we can all understand what actually happened on your side.
> 
> I'll save you the life story, but basically I can't really work on much of anything until at least July 2.
> 
> TLDR: am failing 2 classes, really need to work on school shit

That's ok! feel free to work on this whenever you can :)
